File: frontend/src/services/ai/llm.ts
/**
 * LLM provider abstraction (frontend).
 *
 * Calls the BACKEND AI endpoints (not Groq/Gemini directly).
 * The backend holds the GROQ_API_KEY / GEMINI_API_KEY — never
 * expose keys in the browser bundle.
 *
 * Pattern:
 *   promptBuilder → llm.complete(...) → backend /api/v1/ai/...
 *   AI logic lives in this layer; backend is a thin proxy.
 */

import { api } from "@/api/client";

// --- Types ---

export type LLMProviderName = "groq" | "gemini" | "template";

export interface LLMRequest {
  /** System prompt (role + constraints) */
  system: string;
  /** User prompt (the actual task) */
  user: string;
  /** 0.0-1.0, lower = more deterministic */
  temperature?: number;
  /** Token cap (1 token ≈ 4 chars in Bahasa Indonesia) */
  maxTokens?: number;
  /** Force JSON-mode response (provider must support) */
  jsonMode?: boolean;
  /** Provider override (otherwise uses default) */
  provider?: LLMProviderName;
  /** Try fallback provider first (e.g. for testing) */
  preferFallback?: boolean;
}

export interface LLMResult {
  content: string;
  provider: LLMProviderName;
  model: string;
  tokensUsed: number;
  error: string | null;
}

export interface LLMProvider {
  name: LLMProviderName;
  model: string;
  complete: (req: LLMRequest) => Promise<LLMResult>;
}

// --- Provider implementations (proxy to backend) ---

class BackendProxyProvider implements LLMProvider {
  name: LLMProviderName;
  model: string;
  private endpoint: string;

  constructor(name: LLMProviderName, model: string, endpoint: string) {
    this.name = name;
    this.model = model;
    this.endpoint = endpoint;
  }

  async complete(req: LLMRequest): Promise<LLMResult> {
    try {
      const { data } = await api.post<{
        content: string;
        provider: LLMProviderName;
        model: string;
        tokens_used: number;
        error: string | null;
      }>(this.endpoint, {
        system: req.system,
        user: req.user,
        temperature: req.temperature ?? 0.3,
        max_tokens: req.maxTokens ?? 2048,
        json_mode: req.jsonMode ?? false,
        provider: req.provider,
        prefer_fallback: req.preferFallback ?? false,
      });
      return {
        content: data.content,
        provider: data.provider,
        model: data.model,
        tokensUsed: data.tokens_used,
        error: data.error,
      };
    } catch (e) {
      return {
        content: "",
        provider: this.name,
        model: this.model,
        tokensUsed: 0,
        error: e instanceof Error ? e.message : "LLM request failed",
      };
    }
  }
}

// --- Provider registry ---

const GROQ_PROVIDER: LLMProvider = new BackendProxyProvider(
  "groq",
  "llama-3.1-70b-versatile",
  "/ai/complete",
);

const GEMINI_PROVIDER: LLMProvider = new BackendProxyProvider(
  "gemini",
  "gemini-1.5-flash",
  "/ai/complete",
);

const TEMPLATE_PROVIDER: LLMProvider = {
  name: "template",
  model: "template-v1",
  async complete(_req: LLMRequest): Promise<LLMResult> {
    // The template provider just returns the prompt as content.
    // The orchestrator (ai-analyzer.ts) interprets this and
    // generates deterministic output.
    return {
      content: "[TEMPLATE_PROVIDER_NOT_USED_DIRECTLY]",
      provider: "template",
      model: "template-v1",
      tokensUsed: 0,
      error: null,
    };
  },
};

export const PROVIDERS: Record<LLMProviderName, LLMProvider> = {
  groq: GROQ_PROVIDER,
  gemini: GEMINI_PROVIDER,
  template: TEMPLATE_PROVIDER,
};

// Touch unused symbols to keep them in the bundle (Vite tree-shaking)
// — they'll be used by /prospects/:id detail page in T5 Group 3.
void PROVIDERS;
void GROQ_PROVIDER;
void GEMINI_PROVIDER;

// --- Public API ---

/**
 * Complete a prompt using the configured LLM provider chain
 * (primary → fallback → template).
 *
 * The chain is implemented on the backend; this just calls it.
 * Returns the LLMResult, never throws (errors in `error` field).
 */
export async function complete(req: LLMRequest): Promise<LLMResult> {
  const provider = req.provider
    ? PROVIDERS[req.provider]
    : PROVIDERS.groq;
  return provider.complete(req);
}

/**
 * Parse JSON from LLM response. Handles ```json code fences
 * and bare JSON. Returns null on parse failure.
 */
export function safeParseJson<T = unknown>(text: string): T | null {
  if (!text) return null;
  // Strip code fences
  const fenceMatch = text.match(
    /```(?:json)?\s*([\s\S]*?)\s*```/,
  );
  const candidate = fenceMatch ? fenceMatch[1] : text.trim();
  // Try to find JSON object/array
  const objStart = candidate.indexOf("{");
  const arrStart = candidate.indexOf("[");
  let start = -1;
  let end = -1;
  if (objStart === -1 && arrStart === -1) return null;
  if (objStart === -1) {
    start = arrStart;
    end = candidate.lastIndexOf("]");
  } else if (arrStart === -1) {
    start = objStart;
    end = candidate.lastIndexOf("}");
  } else if (objStart < arrStart) {
    start = objStart;
    end = candidate.lastIndexOf("}");
  } else {
    start = arrStart;
    end = candidate.lastIndexOf("]");
  }
  if (end <= start) return null;
  try {
    return JSON.parse(candidate.slice(start, end + 1)) as T;
  } catch {
    return null;
  }
}

/**
 * Check if LLM is configured (has backend API key).
 * Returns false if backend would fall back to template.
 */
export async function isLLMAvailable(): Promise<boolean> {
  try {
    const { data } = await api.get<{ available: boolean }>("/ai/status");
    return data.available;
  } catch {
    return false;
  }
}

File: backend/app/services/llm.py
"""
Backend LLM service — Groq + Gemini providers with template fallback.

API key NEVER exposed to frontend. Frontend calls /api/v1/ai/...
endpoints which proxy here.
"""
from __future__ import annotations

import json
import logging
import re
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any

import httpx

from app.core.config import settings

logger = logging.getLogger("clientfinder.llm")


@dataclass
class LLMResult:
    content: str
    provider: str
    model: str
    tokens_used: int = 0
    error: str | None = None
    raw: dict[str, Any] | None = None


class LLMError(Exception):
    pass


class BaseProvider(ABC):
    name: str
    model: str

    def __init__(self, api_key: str, model: str | None = None) -> None:
        self.api_key = api_key
        self.model = model or self.model

    @abstractmethod
    async def complete(
        self,
        *,
        system: str,
        user: str,
        temperature: float = 0.3,
        max_tokens: int = 2048,
        json_mode: bool = False,
    ) -> LLMResult: ...


class GroqProvider(BaseProvider):
    name = "groq"
    model = "llama-3.1-70b-versatile"
    base_url = "https://api.groq.com/openai/v1"

    async def complete(
        self,
        *,
        system: str,
        user: str,
        temperature: float = 0.3,
        max_tokens: int = 2048,
        json_mode: bool = False,
    ) -> LLMResult:
        if not self.api_key or self.api_key.startswith("PLACEHOLDER"):
            raise LLMError("Groq API key not configured")
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }
        payload: dict[str, Any] = {
            "model": self.model,
            "messages": [
                {"role": "system", "content": system},
                {"role": "user", "content": user},
            ],
            "temperature": temperature,
            "max_tokens": max_tokens,
        }
        if json_mode:
            payload["response_format"] = {"type": "json_object"}

        async with httpx.AsyncClient(timeout=60.0) as client:
            resp = await client.post(
                f"{self.base_url}/chat/completions",
                headers=headers,
                json=payload,
            )
        if resp.status_code != 200:
            raise LLMError(
                f"Groq HTTP {resp.status_code}: {resp.text[:300]}"
            )
        data = resp.json()
        try:
            content = data["choices"][0]["message"]["content"]
            tokens = data.get("usage", {}).get("total_tokens", 0)
        except (KeyError, IndexError) as e:
            raise LLMError(f"Malformed Groq response: {e}") from e
        return LLMResult(
            content=content,
            provider=self.name,
            model=self.model,
            tokens_used=tokens,
            raw=data,
        )


class GeminiProvider(BaseProvider):
    name = "gemini"
    model = "gemini-1.5-flash"
    base_url = "https://generativelanguage.googleapis.com/v1beta"

    async def complete(
        self,
        *,
        system: str,
        user: str,
        temperature: float = 0.3,
        max_tokens: int = 2048,
        json_mode: bool = False,
    ) -> LLMResult:
        if not self.api_key or self.api_key.startswith("PLACEHOLDER"):
            raise LLMError("Gemini API key not configured")
        url = f"{self.base_url}/models/{self.model}:generateContent?key={self.api_key}"
        payload: dict[str, Any] = {
            "systemInstruction": {"parts": [{"text": system}]},
            "contents": [{"parts": [{"text": user}]}],
            "generationConfig": {
                "temperature": temperature,
                "maxOutputTokens": max_tokens,
            },
        }
        if json_mode:
            payload["generationConfig"]["responseMimeType"] = "application/json"

        async with httpx.AsyncClient(timeout=60.0) as client:
            resp = await client.post(url, json=payload)
        if resp.status_code != 200:
            raise LLMError(
                f"Gemini HTTP {resp.status_code}: {resp.text[:300]}"
            )
        data = resp.json()
        try:
            content = data["candidates"][0]["content"]["parts"][0]["text"]
            tokens = data.get("usageMetadata", {}).get("totalTokenCount", 0)
        except (KeyError, IndexError) as e:
            raise LLMError(f"Malformed Gemini response: {e}") from e
        return LLMResult(
            content=content,
            provider=self.name,
            model=self.model,
            tokens_used=tokens,
            raw=data,
        )


# --- Provider registry ---

def _get_provider(name: str) -> BaseProvider:
    name = name.lower()
    if name == "groq":
        return GroqProvider(
            api_key=settings.llm_primary_api_key,
            model=settings.llm_primary_model,
        )
    if name == "gemini":
        return GeminiProvider(
            api_key=settings.llm_fallback_api_key,
            model=settings.llm_fallback_model,
        )
    raise LLMError(f"Unknown LLM provider: {name}")


async def complete(
    *,
    system: str,
    user: str,
    temperature: float = 0.3,
    max_tokens: int = 2048,
    json_mode: bool = False,
    prefer_fallback: bool = False,
) -> LLMResult:
    """
    Try primary provider, then fallback. Returns the first success.
    Raises LLMError if all providers fail (caller can fall back to template).
    """
    order = (
        [settings.llm_fallback_provider, settings.llm_primary_provider]
        if prefer_fallback
        else [settings.llm_primary_provider, settings.llm_fallback_provider]
    )
    last_err: str | None = None
    for name in order:
        try:
            provider = _get_provider(name)
            return await provider.complete(
                system=system,
                user=user,
                temperature=temperature,
                max_tokens=max_tokens,
                json_mode=json_mode,
            )
        except LLMError as e:
            logger.warning("Provider %s failed: %s", name, e)
            last_err = str(e)
    raise LLMError(f"All providers failed. Last: {last_err}")


def safe_parse_json(text: str) -> dict | list | None:
    """
    Extract JSON from LLM response. Handles ```json code blocks
    and bare JSON. Returns None if parsing fails.
    """
    if not text:
        return None
    # Strip code fences
    fence_match = re.search(
        r"```(?:json)?\s*(\{.*?\}|\[.*?\])\s*```", text, re.DOTALL
    )
    if fence_match:
        text = fence_match.group(1)
    text = text.strip()
    # Try to find a JSON object/array within
    for start_ch, end_ch in [("{", "}"), ("[", "]")]:
        start = text.find(start_ch)
        if start == -1:
            continue
        end = text.rfind(end_ch)
        if end <= start:
            continue
        candidate = text[start : end + 1]
        try:
            return json.loads(candidate)
        except json.JSONDecodeError:
            continue
    # Last resort: try the whole string
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        return None
